handler.go

```diff
@@ -66,45 +66,32 @@ func (h *handler) getUserContexts(reqURL []string, req *http.Request) ([]reflect
 
 	req.ParseForm()
 	for i, userContext := range h.userContexts {
-		if userContext.IsCookies {
-			userContexts[i] = reflect.ValueOf(&Cookies{req: req})
-		} else if userContext.IsHeaders {
-			userContexts[i] = reflect.ValueOf(&Headers{header: req.Header})
-		} else if userContext.ExpectJSONRequest {
-			context, err := generateContext(
-				userContext,
-				newRouteFiller(
-					h.routes,
-					reqURL,
-					userContext.RouteParams,
-					h.matchedPathIndex,
-					h.matchedPath,
-				),
-				newQueryFiller(userContext.QueryParams, req.URL.Query()),
-				newJSONFiller(req.Body),
-			)
-			if err != nil {
-				return nil, err
-			}
-			userContexts[i] = context
+		basicChain := newPipeline().
```

Comment:
I think we should stick in the `generateContext` then we can achieve the same goal without chain.

Reply:
The reason I fall back to a pipeline is that implementation duplicate at two places.

But I have a new idea is to use primitive `append` to combine the whole chain and we use generator function at the end.